core/includes/common.inc

```diff
@@ -1450,32 +1450,59 @@ function valid_url($url, $absolute = FALSE) {
  *
  * @return bool
  *   TRUE if no step mismatch has occured, or FALSE otherwise.
- *
- * @see http://opensource.apple.com/source/WebCore/WebCore-1298/html/NumberInputType.cpp
  */
 function valid_number_step($value, $step, $offset = 0.0) {
-  $double_value = (double) abs($value - $offset);
-
-  // The fractional part of a double has 53 bits. The greatest number that could
-  // be represented with that is 2^53. If the given value is even bigger than
-  // $step * 2^53, then dividing by $step will result in a very small remainder.
-  // Since that remainder can't even be represented with a single precision
-  // float the following computation of the remainder makes no sense and we can
-  // safely ignore it instead.
-  if ($double_value / pow(2.0, 53) > $step) {
-    return TRUE;
+  // Avoid division by zero.
+  if ($step == 0) {
+    return FALSE;
   }
 
-  // Now compute that remainder of a division by $step.
-  $remainder = (double) abs($double_value - $step * round($double_value / $step));
+  // Consider offset, get absolute values.
+  $value = abs($value - $offset);
+  $step = abs($step);
 
-  // $remainder is a double precision floating point number. Remainders that
-  // can't be represented with single precision floats are acceptable. The
-  // fractional part of a float has 24 bits. That means remainders smaller than
-  // $step * 2^-24 are acceptable.
-  $computed_acceptable_error = (double)($step / pow(2.0, 24));
+  // Value and step are the same, no need for math.
+  if ($value == $step) {
+    return TRUE;
+  }
+
+  // Both numbers are integers - use modulo right away.
+  if ((int) $value == $value && (int) $step == $step && $value < PHP_INT_MAX) {
+    return $value % $step == 0;
+  }
+
+  // Convert from possible exponent notation to decimal strings to get the scale
+  // values. Note: sprintf also rounds values but seems to give more reliable
+  // results than number_format.
+  // "10" is the maximum value allowed by the number widget, so higher scale
```

Comment:
> "10" is the maximum value allowed by the number widget,

I wonder why this is the case here 🤔

Reply:
It's in the [settings form](https://github.com/backdrop/backdrop/blob/c6a99af4723ec52cfb35151af762c3efab85f7b8/core/modules/field/modules/number/number.module#L59)

I couldn't tell why :wink: , but 10 is the max for this settings form item.